src/library/scala/collection/Seq.scala

```diff
@@ -30,7 +30,7 @@ trait Seq[+A]
 
   override def equals(o: scala.Any): Boolean =
     o match {
-      case it: Seq[A] => (it canEqual this) && sameElements(it)
+      case it: Seq[A] => (this eq it) || (it canEqual this) && sameElements(it)
```

Comment:
Would it be more efficient to do this before matching on the type? `this.eq(o.asInstanceOf[AnyRef]) || (o match ...)`

Reply:
I'd say it would be either no-worse or no-different, so I've adopted the suggestion.